### PR TITLE
planner: fixup some bugs with `DEFAULT` expression (#13168)(#13211)(#12550)(#11901)

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 

Cherry-pick #13168 #13211 #12550 #11901

Related PR pingcap/parser#648

These are some bugs with `DEFAULT` expression:

1. When `insert t set c1 = default`, TiDB will be panic. This bugs also occur in `REPLACE` / `UPDATE` / `ON DUPLICATED UPDATE` statement (#13211 #13168). Meanwhile, in these statements, if column `c1` is a generated column, MySQL allow these statements, but TiDB will throw an unexpected error (`field c1 not found`). (#12550)
2. TiDB throw an unexpected error when insert `DEFAULT` value into generated columns. e.g. (#11901)
```sql
create table t (a int default 1, b int generated always as (-a));
insert into t value (a, default); // TiDB error: Disallow assign value to generated column.
```
### What is changed and how it works?

To solve 1st bugs, we split a huge PR into 3 tiny PR (#13211 #12550 #13168), for convenient of reviewing. Both of them are merged into `master` branch. After that,  we cherry pick these 3 tiny PR in one PR.

Also, PR #11901 (the 2nd bug above) is related to `DEFAULT` expression, we cherry pick it in passing.

Conflicting files:

- executor/insert_common.go
- expression/expression.go
- planner/core/logical_plan_builder.go
- planner/core/planbuilder.go

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

 - None

Release note

 - None

### CODE REVIEW

**Comment by Deardrops** [comment]
Wait for merge pingcap/parser#648

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -603,19 +602,18 @@ func (b *PlanBuilder) buildSelection(ctx context.Context, p LogicalPlan, where a
...
 	if tblName.L == "" {
 		tblName = c.TblName
+	} else {
+		tblName = colNameField.Name.Table
 	}
-	if dbName.L == "" {
-		dbName = c.DBName
-	}
-	return colName, origColName, tblName, c.OrigTblName, c.DBName
+	return
```
Comment (djshow832) [inline planner/core/logical_plan_builder.go]: This `return` can be removed.

**Comment by djshow832** [comment]
Any conflicting files?

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (Deardrops) [reply planner/core/logical_plan_builder.go]: In Golang, if function has return values, it **must** explicit call `return` keyword. e.g
```golang
func f1() {
	fmt.Println("f1")
}
// print: f1
```
vs
```
func f1() (value int){
	fmt.Println("f2")
	value = 1
	return
}
// compiler error: missing return at end of function
```

**Comment by djshow832** [comment]
LGTM

**Comment by zz-jason** [comment]
@Deardrops It's better to cherry pick these PRs one by one.

**Comment by Deardrops** [comment]
> @Deardrops It's better to cherry pick these PRs one by one.

@zz-jason Got it, I will pay attention to it in later PR.

**Comment by jackysp** [comment]
Please resolve conflicts @Deardrops

**Comment by Deardrops** [comment]
@jackysp all tests passed, PTAL

**Pull request merged**